lib/global-styles.php

```diff
@@ -147,155 +147,325 @@ function gutenberg_experimental_global_styles_get_user_cpt_id() {
+		),
+		'core/heading/h2' => array(
+			'selector' => 'h2',
+			'supports' => array( 'line-height', 'font-size', 'color' ),
+		),
+		'core/heading/h3' => array(
+			'selector' => 'h3',
+			'supports' => array( 'line-height', 'font-size', 'color' ),
+		),
+		'core/heading/h4' => array(
+			'selector' => 'h4',
+			'supports' => array( 'line-height', 'font-size', 'color' ),
+		),
+		'core/heading/h5' => array(
+			'selector' => 'h5',
+			'supports' => array( 'line-height', 'font-size', 'color' ),
+		),
+		'core/heading/h6' => array(
+			'selector' => 'h6',
+			'supports' => array( 'line-height', 'font-size', 'color' ),
+		),
+		'core/columns'    => array(
+			'selector' => '.wp-block-columns',
+			'supports' => array( 'color' ),
+		),
+		'core/group'      => array(
+			'selector' => '.wp-block-group',
+			'supports' => array( 'color' ),
+		),
+		'core/media-text' => array(
+			'selector' => '.wp-block-media-text',
+			'supports' => array( 'color' ),
 		),
-		$theme_supports
 	);
+
+	return $block_data;
+}
+
+/**
+ * Takes a global styles tree and generates
+ * the corresponding stylesheet.
+ *
+ * @param array $tree Global Styles tree.
+ *
+ * @return string
+ */
+function gutenberg_experimental_global_styles_resolver( $tree ) {
+	$stylesheet = '';
+	$block_data = gutenberg_experimental_global_styles_get_block_data();
+	foreach ( array_keys( $tree ) as $block_name ) {
+		if (
+			! array_key_exists( $block_name, $block_data ) ||
+			! array_key_exists( 'selector', $block_data[ $block_name ] ) ||
+			! array_key_exists( 'supports', $block_data[ $block_name ] )
+		) {
+			// Skip blocks that haven't declared support,
+			// so we don't know to process.
+			continue;
+		}
+
+		$token         = '--';
+		$prefix        = '--wp--preset' . $token;
+		$css_variables = gutenberg_experimental_global_styles_get_css_vars( $tree[ $block_name ]['presets'], $prefix, $token );
+
+		$stylesheet .= gutenberg_experimental_global_styles_resolver_styles(
+			$block_data[ $block_name ]['selector'],
+			$block_data[ $block_name ]['supports'],
+			array_merge( $tree[ $block_name ]['styles'], $css_variables )
+		);
+	}
+	return $stylesheet;
+}
+
+/**
+ * Takes a tree and returns the CSS rule
+ * with the proper declarations.
+ *
+ * @param string $block_selector CSS selector for the block.
+ * @param array  $block_supports A list of properties supported by the block.
+ * @param array  $block_styles The list of properties/values to be converted to CSS.
+ *
+ * @return string The corresponding CSS rule.
+ */
+function gutenberg_experimental_global_styles_resolver_styles( $block_selector, $block_supports, $block_styles ) {
+	$css_rule         = '';
+	$css_declarations = '';
+	foreach ( $block_styles as $property => $value ) {
+		// Only convert to CSS:
+		//
+		// 1) The style attributes the block has declared support for.
+		// 2) Any CSS custom property attached to the node.
+		if ( in_array( $property, $block_supports, true ) || strstr( $property, '--' ) ) {
```

Comment:
This part makes sure that we only output the CSS declarations for properties that adhere to the schema (see `gutenberg_experimental_global_styles_get_block_data`).

Example: the heading block supports line-height, font-size, and color. If a theme declared something like:

```
{
    "core/heading/h1": {
        "styles": {
            "color": "var(--wp--preset--color--primary)",
            "font-size": "calc(1px * var(--wp--preset--font-size--huge))",
            "text-shadow": "2px 2px #FF0000"
        }
    }
}
```

The output will only contain the properties valid as per the schema:

```css
h1 {
    color: var(--wp--preset--color--primary);
    font-size: calc(1px * var(--wp--preset--font-size--huge));
}
```

Reply:
I'm not sure if we should filter for the properties the block explicitly supports. It seems it would be nice if we just allowed every property.
What is the plan to support properties that are them selfs variables e.g: a link color. Should we treat them like other CSS properties? And just have something like:
```
    "core/heading/h1": {
        "styles": {
            "color": "var(--wp--preset--color--primary)",
            "font-size": "calc(1px * var(--wp--preset--font-size--huge))",
            "text-shadow": "2px 2px #FF0000"
			"--wp--link-color": '#ff0000',
        }
    }
```

Reply:
> I'm not sure if we should filter for the properties the block explicitly supports. It seems it would be nice if we just allowed every property.

That was initially my thinking as well, but the feedback I've got goes in the other direction. I also think that is better to start with a restricted approach that we then open if there's interest than the other way around (wich can introduce backcompat issues if people rely on that behavior).

>  What is the plan to support properties that are them selfs variables e.g: a link color. Should we treat them like other CSS properties? And just have something like:

That's a good question that should be answered in a different PR. What you suggest should be fine. Depending on how it's done (block support, etc) we can also consider different alternatives.